### PR TITLE
Update namespace restriction

### PR DESCRIPTION
**Pull request by benoitrazet**
This PR contains two commits:
- update the integration test for namespace restriction
- add documentation about namespace restriction

### CODE REVIEW

File: docs/source/architecture/namespace_permission.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  sawtooth.validator.transaction_families='[
+    {\"family\": \"block_info\", \"version\": \"1.0\", \"namespaces\": [\"00b10c\"]},
+    {\"family\":\"sawtooth_settings\", \"version\":\"1.0\"},
+    {\"family\": \"intkey\", \"version\": \"1.0\", \"namespaces\": [\"1cf126\"]} ]'
+
+In the case where a family has no ``"namespace"``, the validator does
+not restrict the write permission of the family. In the example above,
+the family `sawtooth_settings` does not indicate any namespace and
+therefore the validator will let it write at any address.
```
Comment (vaporos) [inline docs/source/architecture/namespace_permission.rst]: ... to any address defined by the intersection of by the namespace(s) present in the transaction processor's registration declaration and the output field of the transaction it is processing.

File: docs/source/architecture/namespace_permission.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+***********************
+
+A transaction family is defined by a set of parameters including a
+name, some version numbers, an *apply* function and namespaces. The
+apply function is what defines the behavior of a transaction family
+according to the `get` and `set` operations on addresses of the Global
+State. The goal of having a family declaring namespaces is to shard
+the set of addresses in order to avoid transaction families tampering
+one another's addreses. More precisely, a namespace consists of an
+address prefix, usually 3 bytes.  When desired, namespace
```
[the remark this answers is not in the document]
Comment (benoitrazet) [reply docs/source/architecture/namespace_permission.rst]: Agreed. I removed the sentence.

File: docs/source/architecture/namespace_permission.rst [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+permissioning is enforced by the validators by ensuring that all `set`
+operations performed by a given transaction family write only to
```
[the remark this answers is not in the document]
Comment (benoitrazet) [reply docs/source/architecture/namespace_permission.rst]: Thanks. It makes sense in this context

File: docs/source/architecture/namespace_permission.rst [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+addresses with the family’s specified namespace prefix(es). By default
+and for better flexibility, this namespace permissioning is not
+enforced by the validators, but there are some scenarios where
+enforcement would increase the security of the execution of
+transaction families.  This would be the case in a scenario where the
+deployment of transaction families is open.
```
[the remark this answers is not in the document]
Comment (benoitrazet) [reply docs/source/architecture/namespace_permission.rst]: I changed the sentence mentioning the security and replaced it with "... feature augments enforcement ...".

Thanks for the additional comments, they make me have a broader vision.

File: docs/source/architecture/namespace_permission.rst (hunk 1 above)
Comment (benoitrazet) [reply docs/source/architecture/namespace_permission.rst]: I am not sure about that one. Some somewhat recent experiment make me believe that the transaction processor's registration declaration is not very helpful for the validator. The transaction processor could give some namespace at registration and later validate transactions not in these namespaces. Isn't it?

Separately, should "intersection" be "union"?

File: integration/sawtooth_integration/tests/test_namespace_permission.py [hunk 5]
```diff
@@ -55,6 +60,16 @@ def get_state(address):
     return base64.b64decode(response['data'])
 
 
+def get_state_prefix(prefix):
```
Comment (dcmiddle) [inline integration/sawtooth_integration/tests/test_namespace_permission.py]: consider clearer method name: get_state_by_prefix

Comment (benoitrazet) [reply integration/sawtooth_integration/tests/test_namespace_permission.py]: I made the change. Thanks!

**Comment by benoitrazet** [comment]
@vaporos @dcmiddleI was thinking for this PR, maybe `namespace protection` would be a better name than `namespace permission` because `permission` is already a concept for key permissioning. I'd be happy to make the change from `namespace permission` to `namespace protection` or `namespace enforcement` (as @vaporos mentioned in an earlier comment) if that would prevent users to be confused with the already existing transactor/validator key permisioning.

**Comment by chenette** [comment]
@benoitrazet - I like " namespace protection".  I was just about to comment that I didn't like using "permissioning" for this feature because it's used for authorization (keys).  Thanks!

**Comment by benoitrazet** [comment]
@chenette Thanks! I update the PR changing the permission to protection, in both the docs and integration test.

File: docs/source/architecture/namespace_protection.rst [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+done during transaction processor registration.
+
+
+In order to make validators enable namespace protection,
+appropriate settings must be published on-chain using the
+:doc:`Settings Transaction Family
+<../transaction_family_specifications/settings_transaction_family>`.
+For every block, a validator loads the json data stored under the
+specific key ``sawtooth.validator.transaction_families``.  At this
+key the data is a list of json entries, one per transaction family,
```
Comment (chenette) [inline docs/source/architecture/namespace_protection.rst]: JSON (same nitpick)

File: docs/source/architecture/namespace_protection.rst [hunk 7]
```diff
@@ -0,0 +1,45 @@
+********************
+Namespace Protection
+********************
+
+A transaction family is defined by a set of parameters including a
+name, some version numbers, an *apply* function and namespaces. The
+apply function is what defines the behavior of a transaction family
+according to the `get` and `set` operations on addresses of the Global
+State. The goal of having a family declaring namespaces is to shard
```
[the remark this answers is not in the document]
Comment (benoitrazet) [reply docs/source/architecture/namespace_protection.rst]: Thanks! I like your last proposition with a longer phrase.

File: docs/source/architecture/namespace_protection.rst [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+protection is not enforced by the validators, but when activated, this
+feature augments enforcement of the transaction processors declaration
+done during transaction processor registration.
+
+
+In order to make validators enable namespace protection,
+appropriate settings must be published on-chain using the
+:doc:`Settings Transaction Family
+<../transaction_family_specifications/settings_transaction_family>`.
+For every block, a validator loads the json data stored under the
```
[the remark this answers is not in the document]
Comment (benoitrazet) [reply docs/source/architecture/namespace_protection.rst]: Thanks!

File: docs/source/architecture/namespace_protection.rst [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    {\"family\": \"block_info\", \"version\": \"1.0\", \"namespaces\": [\"00b10c\"]},
+    {\"family\":\"sawtooth_settings\", \"version\":\"1.0\"},
+    {\"family\": \"intkey\", \"version\": \"1.0\", \"namespaces\": [\"1cf126\"]} ]'
+
+In the case where a family has no ``"namespaces"``, the validator does
+not restrict the write permission of the family. In the example above,
+the family `sawtooth_settings` does not indicate any namespace and
+therefore the validator will let it write at any address. In the case
+of the Settings Transaction Family this is acceptable because it is a
+first class transaction family, that everyone can inspect and review.
```
[the remark this answers is not in the document]
Comment (benoitrazet) [reply docs/source/architecture/namespace_protection.rst]: got rid of "first class transaction family" and replaced it with the notion of transaction families included with Sawtooth. Let me know if it's better now or still unclear.

**Comment by benoitrazet** [comment]
@chenette what would you think of moving this chapter of the documentation to the Section after https://sawtooth.hyperledger.org/docs/core/nightly/master/app_developers_guide/address_and_namespace.html#
I think that might be more appropriate than in the "Architecture" Section as it is proposed in the current PR.

**Comment by benoitrazet** [comment]
The proposed `Namespace Protection` Section is now in the Application Developer's Guide. Because the Section refers to specific details, like `prefix` or `namespace`, it should be put in a place where the reader is expected to have this type of familiarity with Sawtooth.

**Comment by vaporos** [comment]
The text implies is an intended 1-to-1 relationship between namespaces and transaction families. This is not the architectural intent. The intent is that different transaction families can use the same namespaces. Settings is a good example, where we have a "reference implementation" transaction family for updating that state, and other transaction families which read that state as part of their normal operation.

Not sure I like 'protections', but I do agree that we are over-using 'permissions'.  Maybe 'restrictions'? Any other options come to mind?

**Comment by benoitrazet** [comment]
@vaporos I added a couple more sentences in a new commit to make sure the text does not imply a 1-to-1 correspondence (that was not my intent). Let me if you think it is still ambiguous. 

@vaporos @chenette @dcmiddle Concerning `namespace permission` vs `namespace protection` vs `namespace restrictions`, I'm going to add another one in the mix: `namespace enforcement`
Anyone of them is fine with me. Maybe `namespace permission` is the one I prefer the least because of the overloading of permission. Thanks to let me know what you think, I'd be happy to make the change as soon as we reach consensus.

**Comment by benoitrazet** [comment]
@vaporos @chenette @dcmiddle After a good night of sleep, I like `namespace restrictions` because since the feature is solely about the write/set/outputs, as opposed to both inputs and outputs, thus it can be seen as a "restriction".

**Comment by benoitrazet** [comment]
sorry, unexpectedly closed the PR. Reopening it now

**Comment by benoitrazet** [comment]
@vaporos @chenette @dcmiddle I made the change "namespace protection" -> "namespace restriction" in an additional commit.

File: integration/sawtooth_integration/tests/test_namespace_restriction.py [hunk 10]
```diff
@@ -80,26 +95,62 @@ def make_batches(keys):
...
+            'xo take game 5',
+            'xo take game 9',
+            'xo create game',
+            'xo take game 4',
+        ]
 
         # Assert all block info transactions are committed
         for i, batch in enumerate(batches):
             post_batch(batch)
+            send_xo_cmd('{} --url {} --wait {}'.format(
```
Comment (vaporos) [inline integration/sawtooth_integration/tests/test_namespace_restriction.py]: Does this remove the need for the sleep below?  This should probably be in a loop, as the REST API may not wait the total WAIT amount requested.

File: integration/sawtooth_integration/tests/test_namespace_restriction.py [hunk 11]
```diff
@@ -80,26 +95,62 @@ def make_batches(keys):
     return [imf.create_batch([('set', k, 0)]) for k in keys]
 
 
-class TestNamespacePermission(unittest.TestCase):
-    def test_namespace_permission(self):
-        """Tests that namespace permission onchain are letting block_info
-        transactions be injected and preventing intkey transactions to
-        be validated.
+def send_xo_cmd(cmd_str):
+    LOGGER.info('Sending xo cmd')
```
[the remark this answers is not in the document]
Comment (benoitrazet) [reply integration/sawtooth_integration/tests/test_namespace_restriction.py]: Yep, thanks!

File: integration/sawtooth_integration/tests/test_namespace_restriction.py (hunk 10 above)
Comment (benoitrazet) [reply integration/sawtooth_integration/tests/test_namespace_restriction.py]: good catch! It makes sense

**Comment by benoitrazet** [comment]
While I see your point @vaporos that the "namespace restriction" might not be a concern to application developer, I would argue that it depends on the extend of your application. Moreover, as a developer who does not necessarily have control over all the transaction families deployed on a network, it is a legitimate concern that transaction families deployed not tamper the addresses of the transaction family under development. The namespace restriction feature provides an answer to this concern with a single on-chain setting to review.

@vaporos @chenette I could move the documentation in the System's Administrator Guide https://sawtooth.hyperledger.org/docs/core/nightly/master/sysadmin_guide.html after the section "Configuring Permissions". I think it would fit there, and would not bring those concerns to the attention of App developers.

**Comment by chenette** [comment]
@vaporos -  I'm confused by your statement that app developers don't need to know about namespaces.  Isn't that one of the things an app developer needs to design when creating a TP?  And they also need to know that the validator can restrict reads/writes based on other TP namespaces.  To me, this information belongs in the app dev guide.  If it's moved, there should be a subset of this information that's tailored for the app developer.

**Comment by dcmiddle** [comment]
I think the app dev guide is a fine place for this content.
The sysadmin guide, to me, is more about the administration of an individual node.
We could add a section like a _consortium governance guide_ to talk about how to roll out TP upgrades and locking down global settings. For purposes of this PR, though, I'd rather we get the content into the App Dev Guide as it is currently organized.
@chenette if you are ok with that, the lets get your green check and merge this :)

**Pull request merged**